### PR TITLE
reduce number of ES queries to obtain metrics in DOI list for researcher profile

### PR DESCRIPTION
**Pull request by kjgarza**
## Description

Optimized the number of ElasticSearch queries to obtain metrics. 
To avoid 1+N request problems I modified the queries for metrics to request metrics for groups of dois. 
We use very small aggregation per query and the overhead is mostly in the citations query which requires an extra query due to the citations criteria conditions.

related to https://github.com/datacite/bracco/issues/270

## Status
- [x] Ready for Review

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Non Functional Requirement
- [x] Follows the code style of this project.
- [x] Tests Cover Changes
- [ ] All new and existing tests passed.
- [ ] Documentation

### CODE REVIEW

**Comment by kjgarza** [comment]
Some unwanted database changes went through. I'll remove them

File: app/controllers/concerns/metrics_helper.rb [hunk 1]
```diff
@@ -1,17 +1,45 @@
...
-      EventsQuery.new.doi_citations(doi)
+  included do
+
+    def get_metrics_array(dois)
+
+      citations   = EventsQuery.new.citations(dois)
+      views       = EventsQuery.new.views(dois)
+      downloads   = EventsQuery.new.downloads(dois)
+
+      first_merge = merge_array_hashes(citations, views)
```
Comment (mfenner) [inline app/controllers/concerns/metrics_helper.rb]: can this not be done in one step?

File: app/controllers/concerns/metrics_helper.rb [hunk 2]
```diff
@@ -1,17 +1,45 @@
 module MetricsHelper
   extend ActiveSupport::Concern
 
-  class_methods do
-    def doi_citations(doi)
-      EventsQuery.new.doi_citations(doi)
+  included do
+
+    def get_metrics_array(dois)
+
```
Comment (mfenner) [inline app/controllers/concerns/metrics_helper.rb]: linting...

**Comment by kjgarza** [comment]
> I would merge the metrics with the rest of the metadata in the controller, and not the serializer.

@mfenner the problem with that is that all the other metadata comes in a ES:results object merging it with the metrics data would require to transform it into a hash, then that would need to mean we would need to modify the whole Serializer to expect hashes instead of objects like we normally do. Mixing it in the serializer seems like a normal approach https://github.com/Netflix/fast_jsonapi#using-helper-methods

**Comment by mfenner** [comment]
Regarding whether to merge the metrics in the controller or serializer, think about what kind of information it is, for display only (serializer), or more fundamental and maybe also used in other ways. This could for example be benchmarking.

It is easy to generate objects from hashes, e.g. with `OpenStruct`, `Struct` or `Mash` (in the `hashie` gem).

File: app/controllers/concerns/metrics_helper.rb (hunk 1 above)
Comment (kjgarza) [reply app/controllers/concerns/metrics_helper.rb]: changed to one step

File: app/controllers/concerns/metrics_helper.rb (hunk 2 above)
Comment (kjgarza) [reply app/controllers/concerns/metrics_helper.rb]: changed already

**Pull request merged**